Improve 40x pages and http library's handling of them

polybuildr opened the pull request:
Keep just one 40x page and output the error code and message inside
that. Keep a list of codes and messages in http_lib.

Styling of 40x page is currently arbitrary - can be improved when rest
of the site is styled.

Resolves T90.

polybuildr committed:
Improve http error pages

Check if dedicated error page exists and load that. If it doesn't, load
a common error page and output the error code and message inside that.

Styling of error page is currently arbitrary - can be improved when rest
of the site is styled.

polybuildr commented:
{Resolves T90}.

nisargjhaveri commented on src/app/libraries/http_lib.php:
```diff
@@ -8,7 +22,14 @@ function response_code($code, $view_and_exit = true, $view_name = false) {
+            if (isset(self::$httpErrors[$code]['messageDetails'])) {
+                $data['errorMessageDetails'] = self::$httpErrors[$code]['messageDetails'];
+            }
+            $this->load_view('40x', $data);
```
Who said it is 40x?

polybuildr replied:
Good point. Will rename.

nisargjhaveri commented:
Maybe we should do this, 
- load view with exact name if available (i.e. 404). This way we can have a page for 404 and generic page for other 4xx.
- If not, load generic view (i.e. 4xx, 5xx etc.)

The message thing you've done, I'm not so sure about this. I think views should handle it. Pass error code to view. This would be much more extensible. Not so sure, think about it. 

We can/should also allow user to directly pass some data to whatever view is loaded.

polybuildr commented:
That would require a check to see if the view exists, something we certainly don't want to do with a `file_exists`-like function - it would mess up abstraction. If MinimalMVC provided that, we could give it a shot.

Also, at the moment, we don't have any other error pages, so I'll go with this approach (not check if a dedicated view exists). Once we do, we can figure out a way to handle that.

I'll move the messages to the view, no problem. However, one approach to handle the above case is to maintain a list of dedicated error views - which we'd probably have to keep in the library.

polybuildr commented:
Updated PR.
- Updated view name to `http_error`
- Moved message handling to view

nisargjhaveri commented on src/app/views/http_error.php:
```diff
@@ -0,0 +1,88 @@
+        case '404':
+            $message = "Page not found!";
+            $messageDetails = "Sorry, the page you were looking for could not be found.";
+            break;
```
Add default? We can also use original status messages for default.

polybuildr replied:
Original meaning the normal browser standards?

polybuildr commented:
@nisargjhaveri told me that MinimalMVC does handle this and will `return false` if the view does not exist. I have this taken care of locally, but in order for me to update this PR, a small bug has to be taken care of in MinimalMVC.

polybuildr commented:
Updated PR.
- Check if a dedicated view exists, and if not, load the common view.
- Also added a CSS media query to the page and made it look nicer on smaller screens.

nisargjhaveri commented on src/app/libraries/http_lib.php:
```diff
@@ -1,14 +1,19 @@
             }
-            $this->load_view($view_name);
+            // Check if a dedicated view exists; if not, load common error view
+            $dedicatedView = $this->load_view($code);
```
Replace `$code` with `$view_name`. 

We're not using camrlCase in php. 

Also pass the error code in the same way as we do for generic view.

polybuildr replied:
- Good catch. Fixed that.
- Yes, now that that's agreed upon, changed that too.
- What? Why? It's a "dedicated view", why would it need to know the error code?

nisargjhaveri commented on src/app/libraries/http_lib.php:
```diff
@@ -1,14 +1,19 @@
+            // Check if a dedicated view exists; if not, load common error view
+            $dedicatedView = $this->load_view($code);
+            if (! $dedicatedView) {
+                $this->load_view('http_error', [
```
I don't like to call it `http_error`. :/
It is supposed to be generic. We can, in theory, and 200 with this method. The chances of us dog that are very less, so think about it. 
I would prefer views like 4xx, 5xx etc. and also don't see any problem in having 2 different views for server errors and http errors.

polybuildr replied:
That doesn't make any sense. Let's even say we do call this with 200. Why in the world would we want to load view and exit? That will only happen in the case of errors, considering we're returning an http code and then exiting.

We'll modify the code to be different for `4xx` and `5xx` when we have different error views for them, though I have no idea why that would happen. Until then, there's really no point.

nisargjhaveri commented on src/app/libraries/http_lib.php:
```diff
@@ -1,14 +1,19 @@
+            $dedicatedView = $this->load_view($code);
+            if (! $dedicatedView) {
+                $this->load_view('http_error', [
+                    'errorCode' => $code,
```
`status_code`? Please? :P

polybuildr replied:
Like I reasoned above, this will only happen in the case of errors. I don't see the need to make it status code.

nisargjhaveri commented on src/app/views/http_error.php:
```diff
@@ -0,0 +1,98 @@
+    switch ($errorCode) {
+        case '400':
+            $message = 'Invalid request!';
+            $messageDetails = 'Your browser sent an invalid request. Please try again.';
```
We're using underscore_case in php.

polybuildr replied:
:+1:

polybuildr commented:
Updated PR. Addressed code style `under_score` issue. Also moved the CSS to another file. Declined other requests by @nisargjhaveri. Ready for review again.

hharchani commented:
- MinimalMVC requires 404 and 403 pages to be as 404.php and 403.php, so please make those views.
- Mention Felicity (and may be IIIT-H) somewhere, if possible.
- Why is the main heading black? Please make it grey.

polybuildr commented:
@hharchani
- Good point. It does require a `404.php` view, though I couldn't find the requirement for a 403. I've submitted nisargjhaveri/minimalMVC#5 but I don't know if it will be accepted.
- That should also be done. Any suggestions?
- It's black because I liked black - why else? :P What grey would you like it to be?

nisargjhaveri commented on src/app/libraries/http_lib.php:
```diff
@@ -1,14 +1,19 @@
             }
-            $this->load_view($view_name);
+            // Check if a dedicated view exists; if not, load common error view
+            $dedicated_view = $this->load_view($view_name);
```
We should pass status code because here we can give view name directly, and that view can be a shared view for more than one errors. It is not always dedicated view.

polybuildr replied:
Done. Kept the name `dedicatedView`, though, it just makes things clearer.

polybuildr commented:
After felicity-iiith/minimalMVC#1 and #61, this should be okay to merge. This is rather important, because people are seeing this error in production.

I should add a mention to Felicity, though...

polybuildr commented:
Added a mention to Felicity by making the link text  `<< Felicity ʼ16 Home`. Also added the utf-8 charset, which I'd forgotten. :P

hharchani commented:
I think you should send this PR to cap-portal branch

polybuildr commented:
Fair enough.

polybuildr commented:
This is merged in the `cap-portal` branch in #63. I'll close this for now. We can reopen when the task of merging both branches comes up.
